Prevent CSRF, Enforce CRUD and Add BGE Warning on BDI Delete Page

jjbennett opened the pull request:
# Critical Changes

# Changes

- We fixed a security issue with the deletion of NPSP Data Import records. After you click Delete All Data Import Records or Delete Imported Data Import Records, you will now be required to confirm deletion of records by clicking a Mass Delete button. Additionally, the Delete permission on the NPSP Data Imports object is now required to delete NPSP Data Import records. For information on deleting NPSP Data Import records, see the [Administrator's Guide to Importing Donor Data](https://powerofus.force.com/articles/Resource/NPSP-Administrators-Guide-to-Importing-Donor-Data).

# Issues Closed

# New Metadata

# Deleted Metadata

jjbennett commented:
**lurch: attach W-026535

LurchTheButler commented:
Tracking <a href="https://foundation.my.salesforce.com/a2x8000000AYCDnAAP" target="_blank">W-026535</a>

jjbennett commented:
@smarovich - Attached are the 2 new messages that will appear. One to warn the users about deleting and one is for when the user does not have delete permissions. Can you please take a look when you have a moment to help with the wording? I wanted to also mention the 10k record limit but couldn't figure out the best way to explain that in a generic way.

![delete bdi perm error](https://user-images.githubusercontent.com/450473/50169089-7f055180-02b2-11e9-9cdd-ba28d11e1ae2.png)
![delete bdi warning](https://user-images.githubusercontent.com/450473/50169090-7f055180-02b2-11e9-8021-7f5c09e71cad.png)

jjbennett commented:
@smarovich - This branch is ready with the changes applied. Thanks again for the help!

smarovich commented:
![mass_delete](https://user-images.githubusercontent.com/18448536/50365395-386d5c80-0529-11e9-84f1-e892bbc92e0b.png)

force2b commented on src/classes/BDI_DataImportDeleteBTN_CTRL.cls:
```diff
@@ -69,19 +96,39 @@ public with sharing class BDI_DataImportDeleteBTN_CTRL {
-            }
+    private PageReference findDIRecords(Boolean isDeleteAll) {
+        if (isDeleteAll) {
+            dataImports = [SELECT Id FROM DataImport__c LIMIT 10000];
```
Rather than hardcode 10000, what about calculating this dynamically. Though theoretically the value should be 10000, this allows the action to evolve if Salesforce ever increases that limit.
`Integer maxRecords = Limits.getLimitDMLRows() - Limits.getDMLRows();`

jjbennett replied:
@force2b - good idea, fixed!

lmeerkatz commented on src/classes/BDI_DataImportDeleteBTN_CTRL.cls:
```diff
@@ -41,23 +41,50 @@ public with sharing class BDI_DataImportDeleteBTN_CTRL {
+    private List<DataImport__c> dataImports;
+
+    /** @description True when user can delete data import records */
+    @TestVisible
```
@TestVisible and public are redundant here. Should this be private?

Could we simplify this as `public static final Boolean canDelete = SObjectType.DataImport__c.isDeletable();` ?

jjbennett replied:
@lmeerkatz - Thanks for the review! It is a private set, but this allows the test to set it to validate what happens when a user does not have permissions without having to muck around with users/profiles in the test.

jjbennett commented:
@rponti-sforg - This is ready for QA.
FYI - I modified the page message component that are used on several of our pages to properly escape the message when we set escMsg to false. Only BDI Delete and the ContactDeleteOverride pages were setting this value to false. I closed the ContactDeleteOverride work item since the fix here also fixed that issue. Please let me know if you have any questions. Thank you!

jjbennett commented:
@smarovich - Release notes look awesome, thanks!
